ocaml/libs/tracing/tracing.ml

```diff
@@ -0,0 +1,81 @@
+(*
+ * Copyright (C) 2023 Cloud Software Group
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+module SpanContext = struct
+  type t = {trace_id: string; span_id: string} [@@deriving rpcty]
+end
+
+module Span = struct
+  type t = {
+      context: SpanContext.t
+    ; parent: t option
+    ; name: string
+    ; begin_time: float
+    ; end_time: float option
+    ; tags: (string * string) list
+  }
+  [@@deriving rpcty]
+
+  let generate_id n = String.init n (fun _ -> "0123456789abcdef".[Random.int 16])
+
+  let start ?(tags = []) ~name ~parent () =
+    let trace_id =
+      match parent with
+      | None ->
+          generate_id 32
+      | Some span_parent ->
+          span_parent.context.trace_id
+    in
+    let span_id = generate_id 16 in
+    let context : SpanContext.t = {trace_id; span_id} in
+    let begin_time = Unix.gettimeofday () in
+    let end_time = None in
+    {context; parent; name; begin_time; end_time; tags}
+
+  let finish ?(tags = []) ~span () =
+    {span with end_time= Some (Unix.gettimeofday ()); tags= span.tags @ tags}
+
+  let to_string s = Rpcmarshal.marshal t.Rpc.Types.ty s |> Jsonrpc.to_string
+
+  let of_string s =
+    Jsonrpc.of_string s
+    |> Rpcmarshal.unmarshal t.Rpc.Types.ty
+    |> Result.to_option
+end
+
+module Spans = struct
+  let lock = Mutex.create ()
+
+  let spans = Hashtbl.create 100
```

Comment:
Note: once we start using the opentelemetry library we can use its Thread local data structure (I haven't checked how efficiently it is implemented): https://ocaml.org/p/opentelemetry/latest/doc/Opentelemetry/Thread_local/index.html

Reply:
CP-42794